File: telemetry/otel.ts
// telemetry/otel.ts — OpenTelemetry ITelemetryProvider adapter
// Optional peer dependency: npm install @opentelemetry/api @opentelemetry/sdk-node
// Without the package installed, this provider silently no-ops.
import { ITelemetryProvider } from '../core/protocols.js'

// Minimal structural types — avoids hard import of the optional OTel package
interface OTelSpan {
  setStatus(status: { code: number; message?: string }): void
  recordException(err: Error): void
  addEvent(name: string, attrs?: Record<string, unknown>): void
  end(): void
}
interface OTelTracer {
  startActiveSpan<T>(name: string, fn: (span: OTelSpan) => Promise<T>): Promise<T>
}
interface OTelCounter { add(delta: number): void }
interface OTelHistogram { record(value: number): void }
interface OTelObservableGauge {} // eslint-disable-line @typescript-eslint/no-empty-object-type
interface OTelMeter {
  createCounter(name: string): OTelCounter
  createHistogram(name: string): OTelHistogram
  createObservableGauge(name: string, cb: (obs: { observe(v: number): void }) => void): OTelObservableGauge
}
interface OTelActiveSpan { addEvent(name: string, attrs?: Record<string, unknown>): void }
interface OTelApi {
  trace: {
    getTracer(name: string): OTelTracer
    getActiveSpan(): OTelActiveSpan | undefined
  }
  metrics: { getMeter(name: string): OTelMeter }
  SpanStatusCode: { OK: number; ERROR: number }
}

async function loadOTelApi(): Promise<OTelApi | null> {
  try {
    // Dynamic require keeps the package optional at compile time
    const mod = await import('@opentelemetry/api' as string) as OTelApi
    return mod
  } catch {
    return null
  }
}

export class OpenTelemetryProvider implements ITelemetryProvider {
  private _otel: OTelApi | null = null
  private readonly _serviceName: string

  constructor(serviceName = 'ooagent') {
    this._serviceName = serviceName
    loadOTelApi().then(api => { this._otel = api }).catch(() => undefined)
  }

  async span<T>(name: string, fn: () => Promise<T>): Promise<T> {
    const otel = this._otel
    if (!otel) return fn()

    const tracer = otel.trace.getTracer(this._serviceName)
    return tracer.startActiveSpan(name, async (span: OTelSpan) => {
      try {
        const result = await fn()
        span.setStatus({ code: otel.SpanStatusCode.OK })
        return result
      } catch (err) {
        span.setStatus({ code: otel.SpanStatusCode.ERROR, message: (err as Error).message })
        span.recordException(err as Error)
        throw err
      } finally {
        span.end()
      }
    })
  }

  counter(name: string, delta = 1): void {
    const otel = this._otel
    if (!otel) return
    otel.metrics.getMeter(this._serviceName).createCounter(name).add(delta)
  }

  gauge(name: string, value: number): void {
    const otel = this._otel
    if (!otel) return
    otel.metrics.getMeter(this._serviceName).createObservableGauge(name, obs => obs.observe(value))
  }

  histogram(name: string, value: number): void {
    const otel = this._otel
    if (!otel) return
    otel.metrics.getMeter(this._serviceName).createHistogram(name).record(value)
  }

  event(name: string, payload: Record<string, unknown>): void {
    const otel = this._otel
    if (!otel) return
    otel.trace.getActiveSpan()?.addEvent(name, payload)
  }
}

File: src/ooagent/telemetry/otel.py
"""telemetry/otel.py — OpenTelemetry ITelemetryProvider adapter.

Optional dependency: pip install opentelemetry-api opentelemetry-sdk
Without the package installed, this provider silently no-ops.

The TS source keeps the OTel package optional by dynamically `import()`-ing
it at construction time and swallowing the failure. Python's `import` is
synchronous, so the equivalent optional-dependency pattern is a plain
try/except ImportError performed once at module load — there is no
async-loading race to reproduce.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import Any, TypeVar

from ooagent.core.protocols import ITelemetryProvider

T = TypeVar("T")

try:
    from opentelemetry import metrics as _otel_metrics
    from opentelemetry import trace as _otel_trace
    from opentelemetry.metrics import Observation as _OTelObservation
    from opentelemetry.trace import Status as _OTelStatus
    from opentelemetry.trace import StatusCode as _OTelStatusCode
except ImportError:  # pragma: no cover - optional dependency not installed
    _otel_trace = None
    _otel_metrics = None
    _OTelObservation = None
    _OTelStatus = None
    _OTelStatusCode = None


class OpenTelemetryProvider(ITelemetryProvider):
    def __init__(self, service_name: str = "ooagent") -> None:
        self._service_name = service_name
        self._available = _otel_trace is not None and _otel_metrics is not None

    async def span(self, name: str, fn: Callable[[], Awaitable[T]]) -> T:
        if not self._available:
            return await fn()

        tracer = _otel_trace.get_tracer(self._service_name)
        with tracer.start_as_current_span(name) as span:
            try:
                result = await fn()
                span.set_status(_OTelStatus(_OTelStatusCode.OK))
                return result
            except Exception as err:
                span.set_status(_OTelStatus(_OTelStatusCode.ERROR, str(err)))
                span.record_exception(err)
                raise

    def counter(self, name: str, delta: float = 1) -> None:
        if not self._available:
            return
        meter = _otel_metrics.get_meter(self._service_name)
        meter.create_counter(name).add(delta)

    def gauge(self, name: str, value: float) -> None:
        if not self._available:
            return
        meter = _otel_metrics.get_meter(self._service_name)

        def _callback(_options: Any) -> list[Any]:
            return [_OTelObservation(value)]

        meter.create_observable_gauge(name, callbacks=[_callback])

    def histogram(self, name: str, value: float) -> None:
        if not self._available:
            return
        meter = _otel_metrics.get_meter(self._service_name)
        meter.create_histogram(name).record(value)

    def event(self, name: str, payload: dict[str, Any]) -> None:
        if not self._available:
            return
        span = _otel_trace.get_current_span()
        span.add_event(name, attributes=payload)
